marlowe-spec-test/src/Marlowe/Spec/Core/Arbitrary.hs

```diff
@@ -365,16 +503,226 @@ genTransactionOutput :: InterpretJsonRequest -> GenT IO TransactionOutput
+          case cont of
+            Close -> pure $ Transaction_ext times [i] ()
+            _ -> do Transaction_ext _ inputs _ <- arbitraryValidStep state cont
+                    pure $ Transaction_ext times (i:inputs) ()
+  where
+    getAction :: Case -> Action
+    getAction (Case a _) = a
+
+arbitraryValidStep state contract =
+  let
+    nextTimeout Close                                    = minTime state
+    nextTimeout (Pay _ _ _ _ continuation)               = nextTimeout continuation
+    nextTimeout (If _ thenContinuation elseContinuation) = maximum' $ nextTimeout <$> [thenContinuation, elseContinuation]
+    nextTimeout (When _ timeout _)                       = timeout
+    nextTimeout (Let _ _ continuation)                   = nextTimeout continuation
+    nextTimeout (Assert _ continuation)                  = nextTimeout continuation
+  in
+    Transaction_ext <$> arbitraryTimeIntervalAfter (maximum' [minTime state, nextTimeout contract]) <*> pure [] <*> pure ()
+ where
+  maximum' = foldl1 Orderings.max
+
+-- | Generate a random path through a contract.
+arbitraryValidInputs :: State_ext ()             -- ^ The state of the contract.
+                     -> Contract                 -- ^ The contract.
+                     -> Gen [Transaction_ext ()] -- ^ Generator for a transaction inputs.
+arbitraryValidInputs _ Close = pure []
+arbitraryValidInputs state contract =
+  do
+    txIn <- arbitraryValidStep state contract
+    case computeTransaction txIn state contract of  -- FIXME: It is tautological to use `computeTransaction` to filter test cases.
+      TransactionError _ -> pure []
+      TransactionOutput (TransactionOutputRecord_ext _ _ txOutState txOutContract _) -> (txIn :) <$> arbitraryValidInputs txOutState txOutContract
+
+-- | Generate a random case, weighted towards different contract constructs.
+arbitraryCaseWeighted :: [(Int, Int, Int, Int, Int, Int)] -- ^ The weights for contract terms.
+                      -> Context                          -- ^ The Marlowe context.
+                      -> InterpretJsonRequest             -- ^ The JSON request interpreter.
+                      -> GenT IO Case                     -- ^ Generator for a case.
+arbitraryCaseWeighted w context interpret =
+  Case <$> genAction interpret <*> arbitraryContractWeighted w context interpret
+
+-- | Generate an arbitrary contract, weighted towards different contract constructs.
+arbitraryContractWeighted :: [(Int, Int, Int, Int, Int, Int)] -- ^ The weights of contract terms, which must eventually include `Close` as a posibility.
+                          -> Context                          -- ^ The Marlowe context.
+                          -> InterpretJsonRequest             -- ^ The JSON request interpreter.
+                          -> GenT IO Contract                 -- ^ Generator for a contract.
+arbitraryContractWeighted ((wClose, wPay, wIf, wWhen, wLet, wAssert) : w) context interpret =
```

Comment:
The `wWhen` should probably be a `(Int, Int, Int)` itself, marking first the probability of selecting a When and second and third the minimum/maximum number of cases. For the single input case that we get a lot of `[]` for valid inputs, I assume that most generated Whens have empty cases.

Reply:
Currently we do `chooseInt (0, length w)` for generating the number of cases, i.e. the number of cases is bound the by the number remaining steps for adding to the contract. The deeper we go in the contract the more empty case lists we will have